rfc-128-continuous-deployment.md

```diff
@@ -0,0 +1,228 @@
+
+> **It is still possible to deploy an app manually.** The new pipeline is just an automation of the manual deployment steps. Rollback is unchanged, and freezing deployments is supported [[1](https://github.com/alphagov/govuk-puppet/blob/c24ff191ce7fc8a38ceb464f0139faba04b9734b/modules/govuk_jenkins/templates/jobs/deploy_app_downstream.yaml.erb#L22)].
+
+We will continue to use the new development and deployment pipeline created for the trials. Much of the functionality is defined in [a new `Deploy_App_Downstream` Jenkins job](https://github.com/alphagov/govuk-puppet/blob/c24ff191ce7fc8a38ceb464f0139faba04b9734b/modules/govuk_jenkins/templates/jobs/deploy_app_downstream.yaml.erb), which works like this:
+
+1. **Integration**
+
+   1. Deployment runs.
+   1. [**new**] Wait for the Smoke tests for the app to pass [[1](https://github.com/alphagov/govuk-puppet/pull/10387)].
+   1. [**new**] Remotely trigger the same deployment in Staging.
+
+1. **Staging**
+
+   1. [**new**] Repeat as for Integration, but to Production.
+
+We propose to apply this automation across our apps. Given the scale of the change, we need to think carefully about how we can make it consistently safe and secure, before we enable CD for an app.
+
+### Safety concerns
+
+The person who merges a PR must be confident that it is safe to do so. As part of the trial of CD, [we added PR templates to each affected app](https://github.com/alphagov/smart-answers/pull/4758), to explicitly warn developers about the automatic deployment.
+
+However, we also want to ensure that a faulty PR merged in haste can be caught before it impacts the live site. There are a number of checks we can perform to help ensure this, without aiming for perfection.
+
+> **We won't check alerts.** At the time of writing, we have no evidence that checking Icinga alerts has any advantage over the checks discussed below. We can investigate such a check if we find evidence [[1](https://github.com/alphagov/govuk-puppet/pull/10597)].
+
+> **We will only consider _live_ code.** Interactions due to manual and scheduled tasks can be checked manually, using a branch deployment. In the worst case, we should have ample time to fix any faults that develop.
+
+#### Check: Internal repo tests pass (i.e. features are tested in detail)
+
+For the purpose of automating deployments:
+
+- We will use code coverage as our indicator of previous, thorough testing. While an app may have a [rigorous testing strategy](https://github.com/alphagov/content-publisher/blob/69bed3535de1e65bda9d5909a3bc56a319e46576/docs/testing-strategy.md), it's hard to know if this has been applied consistently as the app has evolved.
+
+- We will require a minimum level of coverage that is realistic for all of our apps. We will not pursue this once CD is enabled, as doing so could encourage [a lack of rigour in the way we write our tests](https://jeroenmols.com/blog/2017/11/28/coveragproblem/).
```

Comment:
Oops! That link was originally meant to elaborate on < 100% coverage. You're right it's contradictory here.

I did originally see ongoing coverage out-of-scope for this RFC, because the main goal is to make a one-time "switch to automatic". However, an idea that's surfaced since then is to turn the safety checks proposed here into a minimum ongoing testing standard for GOV.UK. This is better than nothing (as is surfacing coverage stats on GitHub), although still a long way from a practical commitment to better testing.

So, I propose that we drop this point and replace it with (roughly) the following consequence:

### Better testing

In order to meet the safety criteria in this RFC, we will need to improve the tests we have around many of our apps. Once we have enabled automatic deployments, there is a risk the tests will degrade over time.

To help prevent this, we will write new a new manual in GOV.UK Developer Docs, which will specify the safety criteria in this RFC as the new minimum level of testing for GOV.UK apps. This means that, in addition to meeting these criteria for the purpose of enabling automatic deployments, we will commit to upholding them on an ongoing basis. It is out-of-scope for this RFC to resolve how we will uphold this commitment once we have made it.

Reply:
I've added a new commit for this: https://github.com/alphagov/govuk-rfcs/pull/128/commits/922261ea3ab198739dc002c5c79e2bae72d1ef6b.